src/lib/device-online-state.ts

```diff
@@ -0,0 +1,306 @@
+				});
+		})
+		.then(pollInterval => {
+			return Math.max(pollInterval, DEFAULT_SUPERVISOR_POLL_INTERVAL);
+		});
+};
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class NoMessagesToConsumeSignal extends Error {}
+
+class DeviceOnlineStateManager extends EventEmitter {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		super();
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch(err => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				this.consume();
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Online);
+	}
+
+	private deviceOffline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Offline);
+	}
+
+	private deviceTimeout(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Timeout);
+	}
+
+	private consume() {
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessageAsync({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: 30, // prevent other consumers seeing the same message (if any) for 30 secs (prevents multiple API agents from processing it)...
+			})
+			.tap((msg: any) => {
+				// no message! this is thrown to restart the consume process...
+				if (!msg || !msg.id) {
+					return Bluebird.delay(1000).thenThrow(
+						new NoMessagesToConsumeSignal(),
+					);
+				}
+			})
+			.then((msg: any) => {
+				const { id, message } = msg;
+
+				return Bluebird.try(() => {
+					const { uuid, nextState } = JSON.parse(message);
+
+					// raise and event for the state change...
+					switch (nextState) {
+						case DeviceOnlineStates.Timeout:
+							this.scheduleChangeOfStateForDevice(
+								uuid,
+								DeviceOnlineStates.Timeout,
+								DeviceOnlineStates.Offline,
+								API_HEARTBEAT_STATE_TIMEOUT_SECONDS, // put the device into a timeout state if it misses it's scheduled heartbeat window... then mark as offline
+							);
+							this.deviceTimeout(uuid);
+							break;
+						case DeviceOnlineStates.Offline:
+							this.deviceOffline(uuid);
+							break;
+						default:
+							// nothing should be raised here...
+							break;
+					}
+				}).then(() =>
+					this.rsmq.deleteMessageAsync({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id: id,
+					}),
+				);
+			})
+			.catch((err: Error | NoMessagesToConsumeSignal) => {
```

Comment:
I think things should still work without overriding the typing, ie
```suggestion
			.catch((err) => {
```